src/value.jl

```diff
@@ -1,145 +1,181 @@
-function CategoricalValue(level::Integer, pool::CategoricalPool{T, R}) where {T, R}
-    return CategoricalValue(convert(R, level), pool)
+# "categorical value" trait
+abstract type CatValueSupport end
+struct IsCatValue <: CatValueSupport end
+struct NotCatValue <: CatValueSupport end
+
+# by default types do not have "categorical value" trait
+iscatvalue(::Type) = NotCatValue
+iscatvalue(::Type{Union{}}) = NotCatValue # otherwise it dispatches to Type{<:...}
+iscatvalue(x::Any) = iscatvalue(typeof(x))
+
+# union of all types that have "categorical value" trait
+const CatValue{R} = Union{CategoricalValue{T, R} where T,
+                          CategoricalString{R}}
+
+# "categorical value" trait implementation for CategoricalValue and CategoricalString
+iscatvalue(::Type{<:CatValue}) = IsCatValue
+valtype(::Type) = Union{} # no value type if not a proper categorical value type
+valtype(::Type{<:CategoricalValue{T}}) where {T} = T
+valtype(::Type{<:CategoricalString}) = String
+
+# integer type of category reference codes for given type
+reftype(::Type) = DefaultRefType
+# TODO reftype(::Type{T}) where {T <: Integer} = T ?
+reftype(::Type{<:CatValue{R}}) where {R} = R
+
+pool(x::CatValue) = x.pool
+level(x::CatValue) = x.level
+
+# extract the type of the original value from array eltype `T`
+unwrap_catvalue_type(::Type{T}) where {T} =
+    unwrap_catvalue_type(iscatvalue(T), T)
+
+function unwrap_catvalue_type(::Type{T}) where T >: Null
+    V = Nulls.T(T)
+    Union{unwrap_catvalue_type(iscatvalue(V), V), Null}
 end
 
-Base.convert(::Type{CategoricalValue{T, R}}, x::CategoricalValue{T, R}) where {T, R <: Integer} = x
-Base.convert(::Type{CategoricalValue{T}}, x::CategoricalValue{T}) where {T} = x
-Base.convert(::Type{CategoricalValue}, x::CategoricalValue) = x
+unwrap_catvalue_type(::Type{Any}) = # to prevent dispatching to T>:Null method
+    unwrap_catvalue_type(NotCatValue, Any)
+unwrap_catvalue_type(::Type{NotCatValue}, ::Type{T}) where {T} = T
+unwrap_catvalue_type(::Type{IsCatValue}, ::Type{T}) where {T} = valtype(T)
 
-Base.convert(::Type{Union{CategoricalValue{T, R}, Null}}, x::CategoricalValue{T, R}) where {T, R <: Integer} = x
-Base.convert(::Type{Union{CategoricalValue{T}, Null}}, x::CategoricalValue{T}) where {T} = x
-Base.convert(::Type{Union{CategoricalValue, Null}}, x::CategoricalValue) = x
+# get default categorical value type given value type `T` and reference type `R`
+function catvalue_type(::Type{T}, ::Type{R}) where {T, R}
+    V = Nulls.T(T)
+    catvalue_type(iscatvalue(V), V, R)
+end
+
+catvalue_type(::Type{IsCatValue}, ::Type{T}, ::Type{R}) where {T, R} =
+    reftype(T) === R ? T : catvalue_type(valtype(T), R)
```

Comment:
I'm still not convinced the check is useful. Cf. https://github.com/JuliaData/CategoricalArrays.jl/pull/87#discussion_r145643980.